src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,503 @@
+                if (iterationSymbol.Locations.Length != 1 ||
+                    !iterationSymbol.Locations[0].IsInSource ||
+                    iterationVariable != iterationSymbol.Locations[0].FindToken(cancellationToken))
+                {
+                    // was a reference to some other variable.
+                    return;
+                }
+            }
+
+            // Make sure we're starting at 0.
+            var initializerValue = semanticModel.GetConstantValue(initializer, cancellationToken);
+            if (!(initializerValue.HasValue && initializerValue.Value is 0))
+            {
+                return;
+            }
+
+            // Make sure we're incrementing by 1.
+            if (stepValueExpressionOpt != null)
+            {
+                var stepValue = semanticModel.GetConstantValue(stepValueExpressionOpt);
+                if (!(stepValue.HasValue && stepValue.Value is 1))
+                {
+                    return;
+                }
+            }
+
+            var collectionType = semanticModel.GetTypeInfo(collectionExpression, cancellationToken);
+            if (collectionType.Type == null && collectionType.Type.TypeKind == TypeKind.Error)
+            {
+                return;
+            }
+
+            var containingType = semanticModel.GetEnclosingNamedType(context.Span.Start, cancellationToken);
+            if (containingType == null)
+            {
+                return;
+            }
+
+            var ienumerableType = semanticModel.Compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T);
+            var ienumeratorType = semanticModel.Compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerator_T);
+
+            // make sure the collection can be iterated.
+            if (!TryGetIterationElementType(
+                    containingType, collectionType.Type,
+                    ienumerableType, ienumeratorType,
+                    out var iterationType))
+            {
+                return;
+            }
+
+            // If the user uses the iteration variable for any other reason, we can't convert this.
+            var bodyStatements = GetBodyStatements(forStatement);
+            foreach (var statement in bodyStatements)
+            {
+                if (iterationVariableIsUsedForMoreThanCollectionIndex(statement))
+                {
+                    return;
+                }
+            }
+
+            // Looks good.  We can convert this.
+            context.RegisterRefactoring(new MyCodeAction(GetTitle(),
+                c => ConvertForToForEachAsync(
+                    document, forStatement, iterationVariable, collectionExpression,
+                    containingType, collectionType.Type, iterationType, c)));
+
+            // local functions
+            bool iterationVariableIsUsedForMoreThanCollectionIndex(SyntaxNode current)
+            {
+                if (syntaxFacts.IsIdentifierName(current))
```

Comment:
isn't current always statement? 

this is definition
protected abstract SyntaxList<TStatementSyntax> GetBodyStatements(TForStatementSyntax forStatement)

Reply:
I see. it recursively dig in to find identifier.

Reply:
this is recursive.  so we're hitting each statement, then walking down into each one.